### DISCUSSION TITLE
Blog comment on Combining LangChain and Weaviate

### DISCUSSION THREAD

**Discussion by giscus** [weaviate/weaviate-io#715] [Docs/blog comments]
# Blog comment on Combining LangChain and Weaviate

LangChain is one of the most exciting new tools in AI. It helps overcome many limitations of LLMs, such as hallucination and limited input lengths.

https://weaviate.io/blog/combining-langchain-and-weaviate

**Comment by QuantumTutor** [comment]
This is remarkable!

**Reply by erika-shorten** [reply]
Thank you! 🙂

**Comment by DayuanJiang** [comment]
The illustration of map_rerank is wrong.   

> The documents are then ranked and the top two are stuffed to the language model to output a single response.  

After reranking, the answers will not be fed back into the LLM; instead, the best answer will be selected and output directly.

You can see the example output of the [document](https://python.langchain.com/en/latest/modules/chains/index_examples/question_answering.html#the-map-rerank-chain).

**Reply by erika-shorten** [reply]
Hi DayuanJiang, 
Thanks for spotting that! It does make sense that the top doc would be the output, rather than passing the top 2 into the language model for summarization. I guess my explanation was combining map reduce and map re-rank into one.

**Comment by srini047** [comment]
I was following the last code snippet to have a QA based interface. But I am facing this issue:

```ValueError: Error during query: [{'locations': [{'column': 19, 'line': 1}], 'message': 'Unknown argument "nearText" on field "PDF" of type "GetObjectsObj". Did you mean "nearVector" or "nearObject"?', 'path': None}]```


Code used is:
```python
qa = ChatVectorDBChain.from_llm(MyOpenAI, vectorstore)

chat_history = []

print("Welcome to the Weaviate ChatVectorDBChain Demo!")
print("Please enter a question or dialogue to get started!")

while True:
    query = input("")
    result = qa({"question": query, "chat_history": chat_history})
    print(result["answer"])
    chat_history = [(query, result["answer"])]
```

Any help to overcome this is appreciated as this is the last part of the application that needs to be finished to publish it.
Thanks in advance.

**Reply by btconway** [reply]
Have you checked your schema to make sure that you have set a vectorizer for your class? I've had this issue and after doing that it was resolved.

**Reply by srini047** [reply]
Thanks! It worked but facing a new error. The updated schema looks like this:

```
{'classes': [{'class': 'PDF',
   'description': 'Interactive PDF content based using Langchain',
   'invertedIndexConfig': {'bm25': {'b': 0.75, 'k1': 1.2},
    'cleanupIntervalSeconds': 60,
    'stopwords': {'additions': None, 'preset': 'en', 'removals': None}},
   'moduleConfig': {'text2vec-openai': {'model': 'ada',
     'modelVersion': '002',
     'type': 'text',
     'vectorizeClassName': True}},
   'properties': [{'dataType': ['number'],
     'indexFilterable': True,
     'indexSearchable': False,
     'moduleConfig': {'text2vec-openai': {'skip': False,
       'vectorizePropertyName': False}},
     'name': 'embedded_values'}],
   'replicationConfig': {'factor': 1},
   'shardingConfig': {'virtualPerPhysical': 128,
    'desiredCount': 1,
    'actualCount': 1,
    'desiredVirtualCount': 128,
    'actualVirtualCount': 128,
    'key': '_id',
    'strategy': 'hash',
    'function': 'murmur3'},
   'vectorIndexConfig': {'skip': False,
    'cleanupIntervalSeconds': 300,
    'maxConnections': 64,
    'efConstruction': 128,
    'ef': -1,
    'dynamicEfMin': 100,
    'dynamicEfMax': 500,
    'dynamicEfFactor': 8,
    'vectorCacheMaxObjects': 1000000000000,
    'flatSearchCutoff': 40000,
    'distance': 'cosine',
    'pq': {'enabled': False,
     'bitCompression': False,
     'segments': 0,
     'centroids': 256,
     'encoder': {'type': 'kmeans', 'distribution': 'log-normal'}}},
   'vectorIndexType': 'hnsw',
   'vectorizer': 'text2vec-openai'}]}
```

But the error I face when I run the above code is: 
```bash
ValueError: Error during query: [{'locations': [{'column': 6, 'line': 1}], 'message': 'explorer: get class: vectorize params: vectorize params: vectorize params: vectorize keywords: remote client vectorize: API Key: no api key found neither in request header: X-Openai-Api-Key nor in environment variable under OPENAI_APIKEY', 'path': ['Get', 'PDF']}]
```

Any work around for this would be helpful.

**Reply by erika-shorten** [reply]
Hi Sriniketh, 
You're getting the above error because you haven't passed in your OpenAI API key. If you're running Weaviate locally, you can add it to your `yaml` file like this:

```
      ENABLE_MODULES: 'text2vec-openai,generative-openai'
      OPENAI_APIKEY: 'sk-OPENAI_APIKEY'  
```

See the full yaml file here: https://github.com/weaviate/recipes/blob/main/integrations/llamaindex/docker-compose.yml

Adding it to the docker-compose file is one way. You can also pass it at query time, documentation here: https://weaviate.io/developers/weaviate/modules/retriever-vectorizer-modules/text2vec-openai#provide-the-api-key

**Reply by parth-patel2023** [reply]
@btconway can you please help me to solve this error after using the above code 

weaviate.exceptions.UnexpectedStatusCodeException: Query was not successful! Unexpected status code: 422, with response body: {'error': [{'message': 'no graphql provider present, this is most likely because no schema is present. Import a schema first!'}]}

should I create a custom schema ?

**Comment by GoMightyAlgorythmGo** [comment]
Combining multible memory systems will be needed, remember perfection is the enemy of the good and 2 imperfect systems crossreferenced rival the accuracy of 1 system that is very accurate but very expensive and non dynamic. Better make mutlible lightweight memorys with different strentght and dynamic ways to retrive things also add some LLMs to some part of it think of searches think from 1. principles and therys of what hypothetically is the best way for different memory retrival from different storrage (yes different strorrage for different purposes even if some memorys are stored twice or linked in different memorys for differetn purposes some more precise some very impresice but giving a "feel for what was said" a intutition so to say that can often be wrong but is still significantly useful in prioriticing what should be more likely first

**Comment by skitsanos** [comment]
It would be nice to also have an example for using HuggingFace...
